sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java

```diff
@@ -0,0 +1,195 @@
+    public static synchronized Push getInstance() {
+        if (sInstance == null) {
+            sInstance = new Push();
+        }
+        return sInstance;
+    }
+
+    @VisibleForTesting
+    static synchronized void unsetInstance() {
+        sInstance = null;
+    }
+
+    /**
+     * Check whether Push service is enabled or not.
+     *
+     * @return <code>true</code> if enabled, <code>false</code> otherwise.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static boolean isEnabled() {
+        return getInstance().isInstanceEnabled();
+    }
+
+    /**
+     * Enable or disable Push service.
+     *
+     * @param enabled <code>true</code> to enable, <code>false</code> to disable.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static void setEnabled(boolean enabled) {
+        getInstance().setInstanceEnabled(enabled);
+    }
+
+    /**
+     * Enqueue a push installation log.
+     *
+     * @param pushToken the push token value
+     */
+    private void enqueuePushInstallationLog(@NonNull String pushToken) {
+        if (isInactive())
+            return;
+        if (mPushTokenSent)
+            return;
+        PushInstallationLog log = new PushInstallationLog();
+        log.setPushToken(pushToken);
+        mChannel.enqueue(log, PUSH_GROUP);
+        mPushTokenSent = true;
+    }
+
+    /**
+     * Handle push token update success.
+     *
+     * @param token the push token value
+     */
+    @VisibleForTesting
+    synchronized void onTokenRefresh(@NonNull String token) {
+        mPushToken = token;
+        MobileCenterLog.debug(LOG_TAG, "Push token: " + mPushToken);
+        PreferencesStorage.putString(PREFERENCE_KEY_PUSH_TOKEN, mPushToken);
+        enqueuePushInstallationLog(mPushToken);
+    }
+
+    private synchronized void updatePushToken() {
+        if (isInactive())
+            return;
+        if (mPushTokenSent)
+            return;
+        if (mPushToken != null) {
+            enqueuePushInstallationLog(mPushToken);
+        } else {
+            String token = FirebaseInstanceId.getInstance().getToken();
+            if (token != null) {
+                onTokenRefresh(token);
+            }
+        }
+    }
+
+    @Override
+    protected String getGroupName() {
+        return PUSH_GROUP;
+    }
+
+    @Override
+    protected String getServiceName() {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    protected String getLoggerTag() {
+        return LOG_TAG;
+    }
+
+    @Override
+    public Map<String, LogFactory> getLogFactories() {
+        return mFactories;
+    }
+
+    @Override
+    public synchronized void onChannelReady(@NonNull Context context, @NonNull Channel channel) {
+        super.onChannelReady(context, channel);
+        updatePushToken();
+    }
+
+    @Override
+    public synchronized void setInstanceEnabled(boolean enabled) {
+        super.setInstanceEnabled(enabled);
+        if (enabled) {
+            updatePushToken();
+        }
```

Comment:
The rule for modules when disabled is to reset any state, delete persisted state also.
Expected behavior when we disable and re-enable is to get token and send token again.